### ISSUE TITLE
Recommended method to get list of all co-authors with posts

### ISSUE THREAD

**Issue by matthewspencer** [Automattic/Co-Authors-Plus#193]
The template function `coauthors_wp_list_authors()` returns an inconsistent number of guest authors. When using the default argument of `'number'  => 20,`, the function returns less than 20 guest authors in most cases. This makes it difficult to paginate and display guest authors consistently.

This count is off because the bad data returned from the `get_terms()` call is removed [here](https://github.com/Automattic/Co-Authors-Plus/blob/master/template-tags.php#L431-L439):

``` php
foreach( $author_terms as $author_term ) {
    // Something's wrong in the state of Denmark
    if ( false === ( $coauthor = $coauthors_plus->get_coauthor_by( 'user_login', $author_term->name ) ) )
        continue;

    $authors[$author_term->name] = $coauthor;

    $authors[$author_term->name]->post_count = $author_term->count;
}
```

And the guest authors without posts are removed [here](https://github.com/Automattic/Co-Authors-Plus/blob/master/template-tags.php#L464-L467):

``` php
if ( $author->post_count == 0 ) {
    if ( ! $args['hide_empty'] )
        $link = $name;
}
```

Similarly, using `WP_Query` to query the `guest-author` post type like the `CoAuthors_WP_List_Table` class returns both guest authors with and without posts. 

It is possible to make a large (expensive) query of all guest authors and cacheing the cleaned up results, but it is not ideal. 

``` php
global $coauthors_plus;

// Get all co-authors (up to 10,000)
$coauthors = new WP_Query( array(
    'posts_per_page' => 10000,
    'post_type' => 'guest-author',
    'fields' => 'ids',
) );

$coauthors_index = array();

foreach ( $coauthors->posts as $coauthor_id ) {

    $coauthor = $coauthors_plus->get_coauthor_by( 'id', $coauthor_id );

    // Skip co-author if user_nicename is empty
    if ( empty( $coauthor->user_nicename ) )
        continue;

    // Get post count for co-author
    $author_posts = new WP_Query(array(
        'posts_per_page' => 1,
        'author_name' => $coauthor->user_nicename,
        'post_type' => 'post',
        'fields' => 'ids',
    ));

    $post_count = $author_posts->found_posts;

    // Skip co-authors without posts
    if ( 0 === $post_count )
        continue;

    // Add co-author to index
    $coauthors_index[] = $coauthor;

}

// Cache co-authors index as long as possible
wp_cache_set( 'coauthors_index', $coauthors_index );
```

Is there a better method for listing guest authors?

**Comment by philipjohn** [comment]
This may be fixed by v3.3, due imminently.

**Comment by stejus** [comment]
Is there an easier way to list all guest authors with their biography and link to their posts? I am able to list the names and photos with the recommended method. How do I add the biography, email etc. to this?
